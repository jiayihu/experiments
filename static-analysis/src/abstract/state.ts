import { Name } from '../syntax';
import { Domain } from './domain';

export class BottomState {
  type: 'BottomState' = 'BottomState';
}

export const bottomState = new BottomState();

export type FState<T> = (name: Name) => T;
export type State<T> = FState<T> | BottomState;

export const isBottomState = <T>(x: State<T>): x is BottomState => {
  return typeof x !== 'function';
};

const getVarNames = <T>(...states: Array<State<T>>): Name[] => {
  const uniqueNames = new Set(states.flatMap(state => getPrivateNames(state)));

  return Array.from(uniqueNames);
};

const getPrivateNames = <T>(state: State<T>): Array<Name> => {
  if (isBottomState(state)) return [];

  return (state as any)(); // Uses hack provided by `createState`
};

// Implementation hack to return the variable names when needed
const createState = <T>(state: FState<T>, names: Name[]): FState<T> => {
  return (name: Name): T => {
    if (name === undefined) return names as any;

    return state(name);
  };
};

export const substState = <T>(state: State<T>) => (y: Name) => (value: T): State<T> => {
  if (isBottomState(state)) return state;

  const names = Array.from(new Set([...getPrivateNames(state), y]));

  return createState((name: Name) => (name === y ? value : state(name)), names);
};

export const initState = <T>(domain: Domain<T>) => (vars: Array<[Name, T]>): FState<T> => {
  const names = vars.map(([name]) => name);

  return createState((name: Name) => {
    const [, value] = vars.find(([x]) => x === name) || [];

    return value !== undefined ? value : domain.top;
  }, names);
};

export const printState = <T>(state: State<T>): string[] => {
  if (isBottomState(state)) return ['Bottom'];

  return getPrivateNames(state).map(name => {
    return `${name} := ${(state(name) as any)['type']}`;
  });
};

export const eq = <T>(x: State<T>) => (y: State<T>): boolean => {
  if (isBottomState(x) || isBottomState(y)) return x === y;

  return getVarNames(x, y).every(name => x(name) === y(name));
};

export const le = <T>(domain: Domain<T>) => (x: State<T>) => (y: State<T>): boolean => {
  if (isBottomState(x)) return true;

  return !isBottomState(y) && getVarNames(x, y).every(name => domain.le(x(name))(y(name)));
};

export const join = <T>(domain: Domain<T>) => (x: State<T>) => (y: State<T>): State<T> => {
  if (isBottomState(x)) return y;
  if (isBottomState(y)) return x;

  return createState((name: Name) => domain.join(x(name))(y(name)), getVarNames(x, y));
};

export const meet = <T>(domain: Domain<T>) => (x: State<T>) => (y: State<T>): State<T> => {
  if (isBottomState(x) || isBottomState(y)) return bottomState;

  const hasBottom = getVarNames(x, y).find(name => {
    return domain.meet(x(name))(y(name)) === domain.bottom;
  });

  if (hasBottom) return bottomState;

  return createState((name: Name) => domain.meet(x(name))(y(name)), getVarNames(x, y));
};

export const widen = <T>(domain: Domain<T>) => (x: State<T>) => (y: State<T>): State<T> => {
  if (isBottomState(x)) return y;
  if (isBottomState(y)) return x;

  return createState((name: Name) => domain.widen(x(name))(y(name)), getVarNames(x, y));
};

export const stateOps = {
  eq,
  le,
  join,
  meet,
  widen,
};
