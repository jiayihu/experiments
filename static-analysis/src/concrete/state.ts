import { Name, Z } from '../syntax';
import { random } from '../utils';

export type State = (name: Name) => Z;

export const initState: (vars: Record<Name, Z>) => State = (vars: Record<Name, Z>) => name => {
  const value = vars[name];

  return value !== undefined ? value : (vars[name] = random());
};
