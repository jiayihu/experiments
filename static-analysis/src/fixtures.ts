import { Comp, Ass, Num, While, Neg, Eq, Var, Mult, Sub, Skip, True, Le, Add, Div } from './syntax';

export const factorial = new Comp(
  new Ass('y', new Num(1)),
  new While(
    new Neg(new Eq(new Var('x'), new Num(1))),
    new Comp(
      new Ass('y', new Mult(new Var('y'), new Var('x'))),
      new Ass('x', new Sub(new Var('x'), new Num(1))),
    ),
  ),
);

export const division = new Comp(
  new Ass('Q', new Num(0)),
  new Comp(
    new Ass('R', new Var('A')),
    new While(
      new Le(new Var('B'), new Var('R')),
      new Comp(
        new Ass('R', new Sub(new Var('R'), new Var('B'))),
        new Ass('Q', new Add(new Var('Q'), new Num(1))),
      ),
    ),
  ),
);

export const divisionByZero = new Ass('x', new Div(new Num(5), new Num(0)));

export const indirectDivByZero = new Comp(
  new Ass('x', new Num(1)),
  new Comp(
    new Ass('x', new Sub(new Var('x'), new Num(1))),
    new Ass('y', new Div(new Num(5), new Var('x'))),
  ),
);

export const whileNotZeroSkip = new While(new Neg(new Eq(new Var('x'), new Num(0))), new Skip());

export const whileTrueSkip = new While(new True(), new Skip());

export const whileTrueIncrement = new While(
  new True(),
  new Ass('x', new Add(new Var('x'), new Num(1))),
);

export const whileNotZeroIncrement = new While(
  new Neg(new Eq(new Var('x'), new Num(0))),
  new Ass('x', new Add(new Var('x'), new Num(1))),
);

export const hundredLoop = new Comp(
  new Ass('A', new Num(0)),
  new Comp(
    new Ass('B', new Num(0)),
    new While(
      new Le(new Var('A'), new Num(100)),
      new Comp(
        new Ass('A', new Add(new Var('A'), new Num(1))),
        new Ass('B', new Add(new Var('B'), new Num(1))),
      ),
    ),
  ),
);

export const fourtyLoop = new Comp(
  new Ass('x', new Num(0)),
  new While(new Le(new Var('x'), new Num(40)), new Ass('x', new Add(new Var('x'), new Num(1)))),
);

export const whileXGeZeroDecrXAndIncrY = new While(
  new Le(new Num(0), new Var('x')),
  new Comp(
    new Ass('x', new Sub(new Var('x'), new Num(1))),
    new Ass('y', new Add(new Var('y'), new Num(1))),
  ),
);
