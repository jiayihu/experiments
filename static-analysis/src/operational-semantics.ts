import { Stm, State, If, Comp, While, Skip } from './syntax';
import { substState, evalAexpr, evalBexpr } from './eval';

export class InterConfig {
  type: 'InterConfig' = 'InterConfig';
  constructor(readonly stm: Stm, readonly state: State) {}
}

export class FinalConfig {
  type: 'FinalConfig' = 'FinalConfig';
  constructor(readonly state: State) {}
}

export type Config = InterConfig | FinalConfig;

const isFinal = (config: Config): config is FinalConfig => {
  return config.type === 'FinalConfig';
};

export const transition = (config: Config): Config => {
  if (isFinal(config)) return config;

  const { stm, state } = config;

  switch (stm.type) {
    case 'Ass': {
      const { name, aexpr } = stm;
      return new FinalConfig(substState(state)(name)(evalAexpr(aexpr)(state)));
    }
    case 'Skip': {
      return new FinalConfig(state);
    }
    case 'Comp': {
      const { stm1, stm2 } = stm;
      const config1 = transition(new InterConfig(stm1, state));

      return isFinal(config1)
        ? new InterConfig(stm2, config1.state)
        : new InterConfig(new Comp(config1.stm, stm2), config1.state);
    }
    case 'If': {
      const { bexpr, stm1, stm2 } = stm;

      return evalBexpr(bexpr)(state) ? new InterConfig(stm1, state) : new InterConfig(stm2, state);
    }
    case 'While': {
      const { bexpr, stm: whileStm } = stm;

      return new InterConfig(
        new If(bexpr, new Comp(whileStm, new While(bexpr, whileStm)), new Skip()),
        state,
      );
    }
  }
};

export const deriveSeq = (config: Config): Config[] => {
  if (isFinal(config)) return [config];

  return [config, ...deriveSeq(transition(config))];
};

export const semantic = (stm: Stm) => (state: State): State => {
  const derivation = deriveSeq(new InterConfig(stm, state));
  const last = derivation[derivation.length - 1];

  return last.state;
};
