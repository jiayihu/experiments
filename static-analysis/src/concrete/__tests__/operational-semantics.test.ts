import { deriveSeq, InterConfig, semantic } from '../operational-semantics';
import { factorial } from './fixtures';
import { initState } from '../state';

describe('operational semantic', () => {
  it('should derive the configuration', () => {
    const derivation = deriveSeq(new InterConfig(factorial, initState({ x: 3 })));

    expect(derivation).toMatchSnapshot();
  });

  it('should return the factorial result', () => {
    const finalState = semantic(factorial)(initState({ x: 3 }));

    expect(finalState('y')).toBe(6);
    expect(finalState('x')).toBe(1);
  });
});
