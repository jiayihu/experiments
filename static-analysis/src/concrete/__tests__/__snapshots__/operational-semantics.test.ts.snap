// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`operational semantic should derive the configuration 1`] = `
Array [
  InterConfig {
    "state": [Function],
    "stm": Comp {
      "stm1": Ass {
        "aexpr": Num {
          "type": "Num",
          "value": 1,
        },
        "name": "y",
        "type": "Ass",
      },
      "stm2": While {
        "bexpr": Neg {
          "type": "Neg",
          "value": Eq {
            "aexpr1": Var {
              "type": "Var",
              "value": "x",
            },
            "aexpr2": Num {
              "type": "Num",
              "value": 1,
            },
            "type": "Eq",
          },
        },
        "stm": Comp {
          "stm1": Ass {
            "aexpr": Mult {
              "aexpr1": Var {
                "type": "Var",
                "value": "y",
              },
              "aexpr2": Var {
                "type": "Var",
                "value": "x",
              },
              "type": "Mult",
            },
            "name": "y",
            "type": "Ass",
          },
          "stm2": Ass {
            "aexpr": Sub {
              "aexpr1": Var {
                "type": "Var",
                "value": "x",
              },
              "aexpr2": Num {
                "type": "Num",
                "value": 1,
              },
              "type": "Sub",
            },
            "name": "x",
            "type": "Ass",
          },
          "type": "Comp",
        },
        "type": "While",
      },
      "type": "Comp",
    },
    "type": "InterConfig",
  },
  InterConfig {
    "state": [Function],
    "stm": While {
      "bexpr": Neg {
        "type": "Neg",
        "value": Eq {
          "aexpr1": Var {
            "type": "Var",
            "value": "x",
          },
          "aexpr2": Num {
            "type": "Num",
            "value": 1,
          },
          "type": "Eq",
        },
      },
      "stm": Comp {
        "stm1": Ass {
          "aexpr": Mult {
            "aexpr1": Var {
              "type": "Var",
              "value": "y",
            },
            "aexpr2": Var {
              "type": "Var",
              "value": "x",
            },
            "type": "Mult",
          },
          "name": "y",
          "type": "Ass",
        },
        "stm2": Ass {
          "aexpr": Sub {
            "aexpr1": Var {
              "type": "Var",
              "value": "x",
            },
            "aexpr2": Num {
              "type": "Num",
              "value": 1,
            },
            "type": "Sub",
          },
          "name": "x",
          "type": "Ass",
        },
        "type": "Comp",
      },
      "type": "While",
    },
    "type": "InterConfig",
  },
  InterConfig {
    "state": [Function],
    "stm": If {
      "bexpr": Neg {
        "type": "Neg",
        "value": Eq {
          "aexpr1": Var {
            "type": "Var",
            "value": "x",
          },
          "aexpr2": Num {
            "type": "Num",
            "value": 1,
          },
          "type": "Eq",
        },
      },
      "stm1": Comp {
        "stm1": Comp {
          "stm1": Ass {
            "aexpr": Mult {
              "aexpr1": Var {
                "type": "Var",
                "value": "y",
              },
              "aexpr2": Var {
                "type": "Var",
                "value": "x",
              },
              "type": "Mult",
            },
            "name": "y",
            "type": "Ass",
          },
          "stm2": Ass {
            "aexpr": Sub {
              "aexpr1": Var {
                "type": "Var",
                "value": "x",
              },
              "aexpr2": Num {
                "type": "Num",
                "value": 1,
              },
              "type": "Sub",
            },
            "name": "x",
            "type": "Ass",
          },
          "type": "Comp",
        },
        "stm2": While {
          "bexpr": Neg {
            "type": "Neg",
            "value": Eq {
              "aexpr1": Var {
                "type": "Var",
                "value": "x",
              },
              "aexpr2": Num {
                "type": "Num",
                "value": 1,
              },
              "type": "Eq",
            },
          },
          "stm": Comp {
            "stm1": Ass {
              "aexpr": Mult {
                "aexpr1": Var {
                  "type": "Var",
                  "value": "y",
                },
                "aexpr2": Var {
                  "type": "Var",
                  "value": "x",
                },
                "type": "Mult",
              },
              "name": "y",
              "type": "Ass",
            },
            "stm2": Ass {
              "aexpr": Sub {
                "aexpr1": Var {
                  "type": "Var",
                  "value": "x",
                },
                "aexpr2": Num {
                  "type": "Num",
                  "value": 1,
                },
                "type": "Sub",
              },
              "name": "x",
              "type": "Ass",
            },
            "type": "Comp",
          },
          "type": "While",
        },
        "type": "Comp",
      },
      "stm2": Skip {
        "type": "Skip",
      },
      "type": "If",
    },
    "type": "InterConfig",
  },
  InterConfig {
    "state": [Function],
    "stm": Comp {
      "stm1": Comp {
        "stm1": Ass {
          "aexpr": Mult {
            "aexpr1": Var {
              "type": "Var",
              "value": "y",
            },
            "aexpr2": Var {
              "type": "Var",
              "value": "x",
            },
            "type": "Mult",
          },
          "name": "y",
          "type": "Ass",
        },
        "stm2": Ass {
          "aexpr": Sub {
            "aexpr1": Var {
              "type": "Var",
              "value": "x",
            },
            "aexpr2": Num {
              "type": "Num",
              "value": 1,
            },
            "type": "Sub",
          },
          "name": "x",
          "type": "Ass",
        },
        "type": "Comp",
      },
      "stm2": While {
        "bexpr": Neg {
          "type": "Neg",
          "value": Eq {
            "aexpr1": Var {
              "type": "Var",
              "value": "x",
            },
            "aexpr2": Num {
              "type": "Num",
              "value": 1,
            },
            "type": "Eq",
          },
        },
        "stm": Comp {
          "stm1": Ass {
            "aexpr": Mult {
              "aexpr1": Var {
                "type": "Var",
                "value": "y",
              },
              "aexpr2": Var {
                "type": "Var",
                "value": "x",
              },
              "type": "Mult",
            },
            "name": "y",
            "type": "Ass",
          },
          "stm2": Ass {
            "aexpr": Sub {
              "aexpr1": Var {
                "type": "Var",
                "value": "x",
              },
              "aexpr2": Num {
                "type": "Num",
                "value": 1,
              },
              "type": "Sub",
            },
            "name": "x",
            "type": "Ass",
          },
          "type": "Comp",
        },
        "type": "While",
      },
      "type": "Comp",
    },
    "type": "InterConfig",
  },
  InterConfig {
    "state": [Function],
    "stm": Comp {
      "stm1": Ass {
        "aexpr": Sub {
          "aexpr1": Var {
            "type": "Var",
            "value": "x",
          },
          "aexpr2": Num {
            "type": "Num",
            "value": 1,
          },
          "type": "Sub",
        },
        "name": "x",
        "type": "Ass",
      },
      "stm2": While {
        "bexpr": Neg {
          "type": "Neg",
          "value": Eq {
            "aexpr1": Var {
              "type": "Var",
              "value": "x",
            },
            "aexpr2": Num {
              "type": "Num",
              "value": 1,
            },
            "type": "Eq",
          },
        },
        "stm": Comp {
          "stm1": Ass {
            "aexpr": Mult {
              "aexpr1": Var {
                "type": "Var",
                "value": "y",
              },
              "aexpr2": Var {
                "type": "Var",
                "value": "x",
              },
              "type": "Mult",
            },
            "name": "y",
            "type": "Ass",
          },
          "stm2": Ass {
            "aexpr": Sub {
              "aexpr1": Var {
                "type": "Var",
                "value": "x",
              },
              "aexpr2": Num {
                "type": "Num",
                "value": 1,
              },
              "type": "Sub",
            },
            "name": "x",
            "type": "Ass",
          },
          "type": "Comp",
        },
        "type": "While",
      },
      "type": "Comp",
    },
    "type": "InterConfig",
  },
  InterConfig {
    "state": [Function],
    "stm": While {
      "bexpr": Neg {
        "type": "Neg",
        "value": Eq {
          "aexpr1": Var {
            "type": "Var",
            "value": "x",
          },
          "aexpr2": Num {
            "type": "Num",
            "value": 1,
          },
          "type": "Eq",
        },
      },
      "stm": Comp {
        "stm1": Ass {
          "aexpr": Mult {
            "aexpr1": Var {
              "type": "Var",
              "value": "y",
            },
            "aexpr2": Var {
              "type": "Var",
              "value": "x",
            },
            "type": "Mult",
          },
          "name": "y",
          "type": "Ass",
        },
        "stm2": Ass {
          "aexpr": Sub {
            "aexpr1": Var {
              "type": "Var",
              "value": "x",
            },
            "aexpr2": Num {
              "type": "Num",
              "value": 1,
            },
            "type": "Sub",
          },
          "name": "x",
          "type": "Ass",
        },
        "type": "Comp",
      },
      "type": "While",
    },
    "type": "InterConfig",
  },
  InterConfig {
    "state": [Function],
    "stm": If {
      "bexpr": Neg {
        "type": "Neg",
        "value": Eq {
          "aexpr1": Var {
            "type": "Var",
            "value": "x",
          },
          "aexpr2": Num {
            "type": "Num",
            "value": 1,
          },
          "type": "Eq",
        },
      },
      "stm1": Comp {
        "stm1": Comp {
          "stm1": Ass {
            "aexpr": Mult {
              "aexpr1": Var {
                "type": "Var",
                "value": "y",
              },
              "aexpr2": Var {
                "type": "Var",
                "value": "x",
              },
              "type": "Mult",
            },
            "name": "y",
            "type": "Ass",
          },
          "stm2": Ass {
            "aexpr": Sub {
              "aexpr1": Var {
                "type": "Var",
                "value": "x",
              },
              "aexpr2": Num {
                "type": "Num",
                "value": 1,
              },
              "type": "Sub",
            },
            "name": "x",
            "type": "Ass",
          },
          "type": "Comp",
        },
        "stm2": While {
          "bexpr": Neg {
            "type": "Neg",
            "value": Eq {
              "aexpr1": Var {
                "type": "Var",
                "value": "x",
              },
              "aexpr2": Num {
                "type": "Num",
                "value": 1,
              },
              "type": "Eq",
            },
          },
          "stm": Comp {
            "stm1": Ass {
              "aexpr": Mult {
                "aexpr1": Var {
                  "type": "Var",
                  "value": "y",
                },
                "aexpr2": Var {
                  "type": "Var",
                  "value": "x",
                },
                "type": "Mult",
              },
              "name": "y",
              "type": "Ass",
            },
            "stm2": Ass {
              "aexpr": Sub {
                "aexpr1": Var {
                  "type": "Var",
                  "value": "x",
                },
                "aexpr2": Num {
                  "type": "Num",
                  "value": 1,
                },
                "type": "Sub",
              },
              "name": "x",
              "type": "Ass",
            },
            "type": "Comp",
          },
          "type": "While",
        },
        "type": "Comp",
      },
      "stm2": Skip {
        "type": "Skip",
      },
      "type": "If",
    },
    "type": "InterConfig",
  },
  InterConfig {
    "state": [Function],
    "stm": Comp {
      "stm1": Comp {
        "stm1": Ass {
          "aexpr": Mult {
            "aexpr1": Var {
              "type": "Var",
              "value": "y",
            },
            "aexpr2": Var {
              "type": "Var",
              "value": "x",
            },
            "type": "Mult",
          },
          "name": "y",
          "type": "Ass",
        },
        "stm2": Ass {
          "aexpr": Sub {
            "aexpr1": Var {
              "type": "Var",
              "value": "x",
            },
            "aexpr2": Num {
              "type": "Num",
              "value": 1,
            },
            "type": "Sub",
          },
          "name": "x",
          "type": "Ass",
        },
        "type": "Comp",
      },
      "stm2": While {
        "bexpr": Neg {
          "type": "Neg",
          "value": Eq {
            "aexpr1": Var {
              "type": "Var",
              "value": "x",
            },
            "aexpr2": Num {
              "type": "Num",
              "value": 1,
            },
            "type": "Eq",
          },
        },
        "stm": Comp {
          "stm1": Ass {
            "aexpr": Mult {
              "aexpr1": Var {
                "type": "Var",
                "value": "y",
              },
              "aexpr2": Var {
                "type": "Var",
                "value": "x",
              },
              "type": "Mult",
            },
            "name": "y",
            "type": "Ass",
          },
          "stm2": Ass {
            "aexpr": Sub {
              "aexpr1": Var {
                "type": "Var",
                "value": "x",
              },
              "aexpr2": Num {
                "type": "Num",
                "value": 1,
              },
              "type": "Sub",
            },
            "name": "x",
            "type": "Ass",
          },
          "type": "Comp",
        },
        "type": "While",
      },
      "type": "Comp",
    },
    "type": "InterConfig",
  },
  InterConfig {
    "state": [Function],
    "stm": Comp {
      "stm1": Ass {
        "aexpr": Sub {
          "aexpr1": Var {
            "type": "Var",
            "value": "x",
          },
          "aexpr2": Num {
            "type": "Num",
            "value": 1,
          },
          "type": "Sub",
        },
        "name": "x",
        "type": "Ass",
      },
      "stm2": While {
        "bexpr": Neg {
          "type": "Neg",
          "value": Eq {
            "aexpr1": Var {
              "type": "Var",
              "value": "x",
            },
            "aexpr2": Num {
              "type": "Num",
              "value": 1,
            },
            "type": "Eq",
          },
        },
        "stm": Comp {
          "stm1": Ass {
            "aexpr": Mult {
              "aexpr1": Var {
                "type": "Var",
                "value": "y",
              },
              "aexpr2": Var {
                "type": "Var",
                "value": "x",
              },
              "type": "Mult",
            },
            "name": "y",
            "type": "Ass",
          },
          "stm2": Ass {
            "aexpr": Sub {
              "aexpr1": Var {
                "type": "Var",
                "value": "x",
              },
              "aexpr2": Num {
                "type": "Num",
                "value": 1,
              },
              "type": "Sub",
            },
            "name": "x",
            "type": "Ass",
          },
          "type": "Comp",
        },
        "type": "While",
      },
      "type": "Comp",
    },
    "type": "InterConfig",
  },
  InterConfig {
    "state": [Function],
    "stm": While {
      "bexpr": Neg {
        "type": "Neg",
        "value": Eq {
          "aexpr1": Var {
            "type": "Var",
            "value": "x",
          },
          "aexpr2": Num {
            "type": "Num",
            "value": 1,
          },
          "type": "Eq",
        },
      },
      "stm": Comp {
        "stm1": Ass {
          "aexpr": Mult {
            "aexpr1": Var {
              "type": "Var",
              "value": "y",
            },
            "aexpr2": Var {
              "type": "Var",
              "value": "x",
            },
            "type": "Mult",
          },
          "name": "y",
          "type": "Ass",
        },
        "stm2": Ass {
          "aexpr": Sub {
            "aexpr1": Var {
              "type": "Var",
              "value": "x",
            },
            "aexpr2": Num {
              "type": "Num",
              "value": 1,
            },
            "type": "Sub",
          },
          "name": "x",
          "type": "Ass",
        },
        "type": "Comp",
      },
      "type": "While",
    },
    "type": "InterConfig",
  },
  InterConfig {
    "state": [Function],
    "stm": If {
      "bexpr": Neg {
        "type": "Neg",
        "value": Eq {
          "aexpr1": Var {
            "type": "Var",
            "value": "x",
          },
          "aexpr2": Num {
            "type": "Num",
            "value": 1,
          },
          "type": "Eq",
        },
      },
      "stm1": Comp {
        "stm1": Comp {
          "stm1": Ass {
            "aexpr": Mult {
              "aexpr1": Var {
                "type": "Var",
                "value": "y",
              },
              "aexpr2": Var {
                "type": "Var",
                "value": "x",
              },
              "type": "Mult",
            },
            "name": "y",
            "type": "Ass",
          },
          "stm2": Ass {
            "aexpr": Sub {
              "aexpr1": Var {
                "type": "Var",
                "value": "x",
              },
              "aexpr2": Num {
                "type": "Num",
                "value": 1,
              },
              "type": "Sub",
            },
            "name": "x",
            "type": "Ass",
          },
          "type": "Comp",
        },
        "stm2": While {
          "bexpr": Neg {
            "type": "Neg",
            "value": Eq {
              "aexpr1": Var {
                "type": "Var",
                "value": "x",
              },
              "aexpr2": Num {
                "type": "Num",
                "value": 1,
              },
              "type": "Eq",
            },
          },
          "stm": Comp {
            "stm1": Ass {
              "aexpr": Mult {
                "aexpr1": Var {
                  "type": "Var",
                  "value": "y",
                },
                "aexpr2": Var {
                  "type": "Var",
                  "value": "x",
                },
                "type": "Mult",
              },
              "name": "y",
              "type": "Ass",
            },
            "stm2": Ass {
              "aexpr": Sub {
                "aexpr1": Var {
                  "type": "Var",
                  "value": "x",
                },
                "aexpr2": Num {
                  "type": "Num",
                  "value": 1,
                },
                "type": "Sub",
              },
              "name": "x",
              "type": "Ass",
            },
            "type": "Comp",
          },
          "type": "While",
        },
        "type": "Comp",
      },
      "stm2": Skip {
        "type": "Skip",
      },
      "type": "If",
    },
    "type": "InterConfig",
  },
  InterConfig {
    "state": [Function],
    "stm": Skip {
      "type": "Skip",
    },
    "type": "InterConfig",
  },
  FinalConfig {
    "state": [Function],
    "type": "FinalConfig",
  },
]
`;
